connexion/apis/flask_api.py

```diff
@@ -136,113 +135,64 @@ def get_response(cls, response, mimetype=None, request=None):
+            content_type=response.content_type,
+            headers=response.headers,
+            body=response.get_data(),
+        )
 
-        flask_response = cls._build_flask_response(mimetype, content_type,
-                                                   headers, status_code, data)
+    @classmethod
+    def _connexion_to_framework_response(cls, response, mimetype, extra_context=None):
+        """ Cast ConnexionResponse to framework response class """
+        flask_response = cls._build_response(
+            mimetype=response.mimetype or mimetype,
+            content_type=response.content_type,
+            headers=response.headers,
+            status_code=response.status_code,
+            data=response.body,
+            extra_context=extra_context,
+            )
 
         return flask_response
 
     @classmethod
-    def _build_flask_response(cls, mimetype=None, content_type=None,
-                              headers=None, status_code=None, data=None):
+    def _build_response(cls, mimetype, content_type=None, headers=None, status_code=None, data=None, extra_context=None):
+        if cls._is_framework_response(data):
+            return flask.current_app.make_response((data, status_code, headers))
+
+        data, status_code = cls._prepare_body_and_status_code(data=data, mimetype=mimetype, status_code=status_code, extra_context=extra_context)
+
         kwargs = {
             'mimetype': mimetype,
             'content_type': content_type,
-            'headers': headers
+            'headers': headers,
+            'response': data,
+            'status': status_code
         }
         kwargs = {k: v for k, v in six.iteritems(kwargs) if v is not None}
-        flask_response = flask.current_app.response_class(**kwargs)  # type: flask.Response
-
-        if status_code is not None:
-            # If we got an enum instead of an int, extract the value.
-            if hasattr(status_code, "value"):
-                status_code = status_code.value
-
-            flask_response.status_code = status_code
-
-        if data is not None and data is not NoContent:
-            data = cls._jsonify_data(data, mimetype)
-            flask_response.set_data(data)
-
-        elif data is NoContent:
-            flask_response.set_data('')
-
-        return flask_response
+        return flask.current_app.response_class(**kwargs)  # type: flask.Response
 
     @classmethod
     def _jsonify_data(cls, data, mimetype):
-        if (isinstance(mimetype, six.string_types) and is_json_mimetype(mimetype)) \
-                or not (isinstance(data, six.binary_type) or isinstance(data, six.text_type)):
+        # TODO: to discuss: Using jsonifier for all type of data, even when mimetype is not json is strange. Why ?
+        if (isinstance(mimetype, str) and is_json_mimetype(mimetype)) \
+                or not (isinstance(data, bytes) or isinstance(data, str)):
             return cls.jsonifier.dumps(data)
```

Comment:
I think connexion started out as a tool focused on json apis, and so there is some very old and very weird coupling with json.
In my opinion, the number one improvement I'd like to see from connexion is to decouple content types (see my work here https://github.com/zalando/connexion/pull/760 ), but it would mean a backwards incompatible change, and so a major version bump.

With respect to this particular section, I agree - it doesn't think it makes any sense to jsonify everything.
My guess though, is that because connexion is so forgiving about not specifying response content type, there are a bunch of users who are not specifying a json content type for their response, and are relying on this (and other) bugs/assumptions to get json out the other end.

I think the only thing to do in the context of this PR is leave it for now, and then if we can muster the effort - work together to decouple content type properly and do a major version bump.

I'm very interested to pick #760 back up, and to collaborate on this, assuming @hjacobs will have enough bandwidth to review the changes and merge them.

Reply:
Regarding what to do about the mismatch in aiohttp vs flask - maybe we can think about it from the point of view of the user. What would be the principle of least surprise?

On the one hand, we value correctness and having the same behavior between aiohttp and flask.
On the other hand, if we change the behavior in a way that isn't backwards compatible, then we should probably either be staging a major version bump, or at least understanding the breakage risk very well.

Maybe one way to understand the risk better would be to come up with an example application that returns different data based on this change - is it hard to trigger? how many people might it affect?

Reply:
Oh. This accomplishes the whole dict => json default. hmmm

Reply:
So, previous to this change no aiohttp users would be able to directly return a dictionary anyway. They would have had to return a ConnexionResponse or a web.StreamResponse. So, enabling passing a dictionary here should not be backwards incompatible. I think I have a change to add this. I'll push it momentarily.

Reply:
oh yeah - you're right it is backwards compatible - nice!

Reply:
this is a tricky one! upon more reflection, if your intent is to allow a seamless switch between aiohttp and flask, then I think it makes sense to adopt the flask tuple behavior into abstract. One way to think of this is "if there's no content-type, then default to using json for serialization". I think that's a reasonable default, and it's much more useful and predictable than str(). If folks want to override the default, they can return a ConnexionResponse, framework response, or specify the content-type explicitly. What do you think?

Reply:
OK. I went ahead and got rid of `flask_api._jsonify_data()`. See the commit message about why I think this is a combination of new feature (`str()` instead of `TypeError`) and bugfix (do not jsonify data with non-json mimetypes). What do you think? Does that make sense?

ie - use the behavior from flask for the blue square, and use the behavior from aiohttp for the yellow squares:

![Selection_033](https://user-images.githubusercontent.com/1558590/70362328-312e8400-184a-11ea-8129-87f19140c6b8.png)

Reply:
parallel conversation going on gitter, but IMO str() is anti-useful. I think it's important for APIs to return structured data by default, and if the data can't be represented in a structured way it's better to error (and require the developer to implement serialization) than call str() and return something unstructured.